### PR TITLE
Measurement Controlled operations

### PR DESCRIPTION
**Pull request by JamesB-1qbit**
The initial version of measurement controlled operations.
Users introduce a "CMEASURE" Gate
Three choices
1) Dictionary for parameter, i.e. `Gate("CMEASURE", target, parameter= {"0": List[Gate], "1": List[Gate]})`
2) Function in the parameter, i.e. `Gate("CMEASURE", target, parameter=function(measurement))`
3) Subclass of ClassicalControl instantiated when initializing the Circuit, i.e. circuit = Circuit(List[Gate], cmeasure_control=Class)

The last is by far the most flexible as can be seen for iterative phase estimation in `test_simulator.py`.

### CODE REVIEW

File: tangelo/linq/circuit.py
Comment (ValentinS4t1qbit) [inline tangelo/linq/circuit.py:172]: I think this is more straightforward : `if CMEASURE in self.counts`

File: tangelo/toolboxes/post_processing/post_selection.py
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/post_processing/post_selection.py:156]: just call call them freqs1 and freqs2. Your second name gives no information and we can clearly see they're dicts.

same for `n_to_keep`: why not just `n` ? your docstring makes it really clear

File: tangelo/toolboxes/post_processing/post_selection.py
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/post_processing/post_selection.py:152]: digits implies more than just 0 or 1, i'd consider "bits" instead.

File: tangelo/linq/target/target_qulacs.py
Comment (ValentinS4t1qbit) [inline tangelo/linq/target/target_qulacs.py:165]: I thinl L165 could be deleted. it recomputes the same thing as L164 and you're not using the variable for the return statement right after.

File: tangelo/linq/target/target_cirq.py
Comment (ValentinS4t1qbit) [inline tangelo/linq/target/target_cirq.py:149]: Lets try to skip at least 1 line and have 1-2 comments for this 15 line blob

**Review by ValentinS4t1qbit — CHANGES_REQUESTED** [review]
First pass.

I think the second pass will focus more on your target_cirq and target_qulacs code, which should be a bit easier to understand for me, and allow me to be more insightful

File: tangelo/linq/target/target_cirq.py
Comment (JamesB-1qbit) [reply tangelo/linq/target/target_cirq.py:119]: In general, we don't know how many iterations we are going through as the length of `unitary_circuits` can increase if the CMEASURE gate returns gates with MEASURE or CMEASURE. For example, one of the tests is
```python3
Circuit([Gate("H", 0),
         Gate("CMEASURE", 0, parameter={"0": [Gate("H", 0), Gate("CMEASURE", 0)], "1": []})])
```
The measurement has a 50 percent probability of either adding a unitary (Gate("H",0)) or continuing. So one could have a circuit that measures "0" multiple times and the length of `unitary_circuits` lengthens multiple times.

File: tangelo/linq/target/target_cirq.py
Comment (JamesB-1qbit) [reply tangelo/linq/target/target_cirq.py:94]: If they use "desired_meas_result", they do not need to specify the number of shots as the circuit is deterministic. If they do not specify the number of shots but have a circuit with measurements, this `ValueError` is caught in the `simulate` function of `Backend`.

File: tangelo/linq/circuit.py
Comment (ValentinS4t1qbit) [inline tangelo/linq/circuit.py:179]: Please use `'CMEASURE' in self.counts` as this is truly what you want here, rather than having a convoluted way to look for the actual count, returning 0 if you cant find it, and using that as a true/false value to signify that `CMEASURE` appeared in the counts.

File: tangelo/linq/circuit.py
Comment (ValentinS4t1qbit) [inline tangelo/linq/circuit.py:415]: For these 2 functions: is it reasonable to expect the user to pass something that is not covered in these if branches ? If yes, how do we handle them ?

Currently, I think both the functions will do nothing and ignore what the user has passed, without any warning etc.

File: tangelo/linq/target/target_stim.py
Comment (ValentinS4t1qbit) [inline tangelo/linq/target/target_stim.py:61]: Return a `NotImplementedError`.

File: tangelo/linq/target/target_qiskit.py
Comment (ValentinS4t1qbit) [inline tangelo/linq/target/target_qiskit.py:77]: Return a `NotImplementedError`.

File: tangelo/linq/target/target_qdk.py
Comment (ValentinS4t1qbit) [inline tangelo/linq/target/target_qdk.py:67]: Return a `NotImplementedError`.

File: tangelo/linq/target/target_sympy.py
Comment (ValentinS4t1qbit) [inline tangelo/linq/target/target_sympy.py:57]: Return a `NotImplementedError`.

File: tangelo/linq/tests/test_simulator.py
Comment (ValentinS4t1qbit) [inline tangelo/linq/tests/test_simulator.py:732]: Why not just setting it to n_bits ? 

You do not need to copy anything here, n_bits (and self.n_bits) is immutable: it's an integer. Even if you changed the value of n_bits later in the body of this function, it wouldn't affect other variables.

```
class xy():
    
    def __init__(self):
        self.x = 1
        self.y = self.x
        
o = xy()

o.x = 2
print(o.x, o.y)   # prints 2 1
```

File: tangelo/linq/tests/test_simulator.py
Comment (ValentinS4t1qbit) [inline tangelo/linq/tests/test_simulator.py:759]: no need for copy. self.n_bits is immutable so you are getting a new variable pointing to that specific value.

File: tangelo/linq/target/target_qulacs.py
Comment (ValentinS4t1qbit) [reply tangelo/linq/target/target_qulacs.py:188]: I do not understand why this was resolved and why L188 couldn't be deleted. The value of that variable is not used in the return statement hit by the code on the next line, and will be lost.

File: tangelo/linq/target/target_cirq.py
Comment (ValentinS4t1qbit) [inline tangelo/linq/target/target_cirq.py:76]: NotImplementedError

**Review by ValentinS4t1qbit — CHANGES_REQUESTED** [review]
We may be done after that second review, I think it reads pretty well and its just minor stuff

**Review by ValentinS4t1qbit — APPROVED** [review]
Thank you James, good work !

**Comment by ValentinS4t1qbit** [comment]
Currently 1 test fails:

```
FAILED linq/tests/test_simulator.py::TestSimulateMisc::test_measurement_controlled_gates_not_supported - ValueError: Gate 'CMEASURE' not supported on backend qdk
```

This suggests two things:

1. We missed a spot that should say `NotImplementedError` in qdk :)
2. Your test currently includes a backend that should not be included

Once we've fixed that, we are ready to merge !

**Review by ValentinS4t1qbit — APPROVED** [review]

**Pull request merged**
